src/Orleans/Core/Grain.cs

```diff
@@ -38,26 +38,51 @@ namespace Orleans
     /// </summary>
     public abstract class Grain : IAddressable
     {
-        internal CodeGeneration.GrainState GrainState { get; set; }
+        private IGrainRuntime runtime;
+        internal IGrainState GrainState { get; set; }
 
+        // Do not use this directly because we currently don't provide a way to inject it;
+        // any interaction with it will result in non unit-testable code. Any behaviour that can be accessed 
+        // from within client code (including subclasses of this class), should be exposed through IGrainRuntime.
+        // The better solution is to refactor this interface and make it injectable through the constructor.
         internal IActivationData Data;
 
-        internal GrainReference GrainReference { get { return Data.GrainReference; } }
+        internal IGrainRuntime Runtime
+        {
+            get { return runtime; }
+            set
+            {
+                runtime = value;
+                GrainFactory = value.GrainFactory;
+            }
+        }
+
+        protected IGrainFactory GrainFactory { get; private set; }
+
+        internal IGrainIdentity Identity;
 
         /// <summary>
-        /// This grain's unique identifier.
+        /// This constructor should never be invoked. We expose it so that client code (subclasses of Grain) do not have to add a constructor.
+        /// Client code should use the GrainFactory property to get a reference to a Grain.
         /// </summary>
-        internal GrainId Identity
+        protected Grain()
         {
-            get { return Data.Identity; }
         }
 
+        protected Grain(IGrainIdentity identity, IGrainRuntime runtime)
```

Comment:
Exactly! And as @sergeybykov explained it to me, we **have to use setter-injection in the runtime**, otherwise ALL grain impl. classes would have to impl. this constructor.